### PR TITLE
WIP: Sending files of Messenger module

### PR DESCRIPTION
**Pull request by NamsooCho**

### CODE REVIEW

File: src/toxcore/net_crypto/mod.rs [hunk 1]
```diff
@@ -763,7 +763,7 @@ impl NetCrypto {
     }
 
     /// Send `CryptoData` packet if the connection is established.
-    fn send_data_packet(&self, connection: &mut CryptoConnection, data: Vec<u8>, packet_number: u32)
+    pub fn send_data_packet(&self, connection: &mut CryptoConnection, data: Vec<u8>, packet_number: u32)
```
Comment (kurnevsky) [inline src/toxcore/net_crypto/mod.rs]: No need to make it public.

File: src/toxcore/messenger/packet/mod.rs [hunk 2]
```diff
@@ -12,7 +12,7 @@ mod msi;
 mod file_control;
 mod typing;
 mod user_status;
-mod file_data;
+pub mod file_data;
```
Comment (kurnevsky) [inline src/toxcore/messenger/packet/mod.rs]: No need to make it public - we have `pub use`.

File: src/toxcore/dht/packet/mod.rs [hunk 3]
```diff
@@ -15,7 +15,7 @@ mod cookie_response;
 mod bootstrap_info;
 mod lan_discovery;
 mod crypto_handshake;
-mod crypto_data;
+pub mod crypto_data;
```
Comment (kurnevsky) [inline src/toxcore/dht/packet/mod.rs]: No need to make it public - we have `pub use`.

File: src/toxcore/net_crypto/crypto_connection.rs [hunk 4]
```diff
@@ -664,6 +672,25 @@ impl CryptoConnection {
...
+
+    /// Check if packet_number was received by the other size.
+    ///
+    /// Note: The condition `buffer_end - buffer_start < packet_number - buffer_start` is
+    /// a trick which handles situations `buffer_end >= buffer_start` and
+    /// buffer_end < buffer_start` (when buffer_end overflowed) both correctly.
+    ///
+    /// It CANNOT be simplified to `packet_number < buffer_start`, as it will fail
+    /// when `buffer_end < buffer_start`.
+    pub fn is_packet_received(&self, packet_number: u32) -> bool {
```
Comment (kurnevsky) [inline src/toxcore/net_crypto/crypto_connection.rs]: Where is it used?

File: src/toxcore/messenger/packet/mod.rs (hunk 2 above)
Comment (NamsooCho) [reply src/toxcore/messenger/packet/mod.rs]: >use crate::toxcore::messenger::packet::file_data::MAX_FILE_DATA_SIZE;

is used.

File: src/toxcore/dht/packet/mod.rs (hunk 3 above)
Comment (NamsooCho) [reply src/toxcore/dht/packet/mod.rs]: >use crate::toxcore::dht::packet::crypto_data::MAX_CRYPTO_DATA_SIZE;

is used.

File: src/toxcore/net_crypto/crypto_connection.rs (hunk 4 above)
Comment (NamsooCho) [reply src/toxcore/net_crypto/crypto_connection.rs]: done

File: src/toxcore/dht/packet/mod.rs (hunk 3 above)
Comment (kpp) [reply src/toxcore/dht/packet/mod.rs]: Have a look at `pub(crate)`

File: src/toxcore/messenger/packet/mod.rs (hunk 2 above)
Comment (NamsooCho) [reply src/toxcore/messenger/packet/mod.rs]: done

File: src/toxcore/dht/packet/mod.rs (hunk 3 above)
Comment (NamsooCho) [reply src/toxcore/dht/packet/mod.rs]: Used ....::packet{Packet as DhtPacket, *};

File: src/toxcore/messenger/packet/file_send_request.rs [hunk 5]
```diff
@@ -79,11 +79,16 @@ Length    | Content
...
-    file_name: String,
+    /// File id used in file transfer session
+    pub file_id: u8,
+    /// Type of transfer file
+    pub file_type: FileType,
+    /// Size of transfer file
+    pub file_size: u64,
+    /// A random key
+    pub file_unique_id: FileUID,
+    /// UTF-8 C string
```
Comment (kpp) [inline src/toxcore/messenger/packet/file_send_request.rs]: Not C string. C strings is to indicate how raw bytes should look like (NUL terminated)

Comment (NamsooCho) [reply src/toxcore/messenger/packet/file_send_request.rs]: done

**Comment by kpp** [comment]
Changes from src/toxcore/messenger/packet/* should be in a separate PR

**Comment by NamsooCho** [comment]
>Changes from src/toxcore/messenger/packet/* should be in a separate PR

done.
But changes need to compile are included.
Other changes will be in other PR.

File: src/toxcore/onion/client/mod.rs [hunk 6]
```diff
@@ -127,6 +129,14 @@ struct OnionFriend {
     search_count: u32,
     /// Time when this friend was seen online last time
     last_seen: Option<Instant>,
+    /// Number of files sending.
+    pub num_sending_files: u8,
+    /// File transfer objects for sending.
+    pub files_sending: Arc<RwLock<Vec<Option<FileTransfers>>>>,
```
Comment (kpp) [inline src/toxcore/onion/client/mod.rs]: In the future we can get rid of Onion, so there should be messenger/friend with these fields like num_sending_files, files_sending, files_receiving, status etc, and with a field with a link to OnionFriend

File: src/toxcore/messenger/packet/file_control.rs [hunk 7]
```diff
@@ -16,6 +16,17 @@ pub enum TransferDirection {
     Receive
 }
 
+impl TransferDirection {
+    /// Toggle direction
+    pub fn toggle(&self) -> Self {
```
Comment (kurnevsky) [inline src/toxcore/messenger/packet/file_control.rs]: It's `Copy` so can be taken by value. btw, doesn't clippy warn here?

File: src/toxcore/onion/client/mod.rs (hunk 6 above)
Comment (NamsooCho) [reply src/toxcore/onion/client/mod.rs]: I think this would be in other PR.

File: src/toxcore/messenger/packet/file_control.rs (hunk 7 above)
Comment (NamsooCho) [reply src/toxcore/messenger/packet/file_control.rs]: Now, there are much warnings when compile. So I didn't check warnings.
All warnings will be removed.

File: src/toxcore/onion/client/mod.rs (hunk 6 above)
Comment (kpp) [reply src/toxcore/onion/client/mod.rs]: OK

File: src/toxcore/messenger/file_sending/packet/mod.rs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+impl FromBytes for FileUID {
+    named!(from_bytes<FileUID>, map_opt!(take!(FILE_UID_BYTES), FileUID::from_slice));
+}
+
+/** File sending packet enum that encapsulates all types of file sending packets.
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Packet {
```
Comment (kpp) [inline src/toxcore/messenger/file_sending/packet/mod.rs]: Make a new PR with file_transfer::Packet changes only

**Comment by NamsooCho** [comment]
This PR is outdated.
I am closing it and will make a new PR soon.

**Pull request closed without merging**